language/diem-vm/parallel-executor/src/outcome_array.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    errors::{Error, Result},
+    task::ExecutionStatus,
+};
+use once_cell::sync::OnceCell;
+
+pub(crate) struct OutcomeArray<T, E> {
+    // Hold the execution results for each individual transactions. Each cell should be set exactly
+    // once.
+    results: Vec<OnceCell<ExecutionStatus<T, Error<E>>>>,
+}
+
+impl<T: Default, E: Send> OutcomeArray<T, E> {
+    pub fn new(len: usize) -> OutcomeArray<T, E> {
+        OutcomeArray {
+            results: (0..len).map(|_| OnceCell::new()).collect(),
+        }
+    }
+
+    pub fn set_result(&self, idx: usize, res: ExecutionStatus<T, Error<E>>) {
+        // Only one thread can write at the time, so just set it.
+
+        let entry = &self.results[idx];
+        assert!(entry.set(res).is_ok());
+    }
+
+    pub fn get_all_results(self, valid_length: usize) -> Result<Vec<T>, E> {
+        let len = self.results.len();
+        let mut results = self
+            .results
+            .into_iter()
+            .take(valid_length)
+            .map(|cell| cell.into_inner())
+            .collect::<Option<Vec<_>>>()
+            .ok_or(Error::InvariantViolation)?;
+
+        match results.pop() {
+            Some(ExecutionStatus::Success(output)) => {
+                // Last transaction is executed successfully, meaning that all transactions outputs will be returned.
+                assert_eq!(len, valid_length);
+                let mut other_outputs = results
+                    .into_iter()
+                    .map(|status| match status {
+                        ExecutionStatus::Success(output) => Ok(output),
+                        _ => Err(Error::InvariantViolation),
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+                other_outputs.push(output);
+                Ok(other_outputs)
+            }
+            Some(ExecutionStatus::SkipRest(output)) => {
+                // Last transaction calls to skip the trailing transactions, thus we will fill the
+                // rest of the transaction outputs with default values
+                let mut other_outputs = results
+                    .into_iter()
+                    .map(|status| match status {
+                        ExecutionStatus::Success(output) => Ok(output),
+                        _ => Err(Error::InvariantViolation),
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+                other_outputs.push(output);
+                other_outputs.resize_with(len, Default::default);
```

Comment:
The reason why I don't want it be `Vec<Option<T>>` is that the caller will need to do another round of `into_iter` and cause another round of allocation&deallocation cost, which could be huge when the vector size is big when it comes to parallel. Was wonder instead of relying on `default`, would another function in the trait to construct the output in this case a better option? Or simply just return a `Vec<T>` and ask the caller to fill in the rest of the outputs?